### DISCUSSION TITLE
How to get a Button's element reference ?

### DISCUSSION THREAD

**Discussion by tmarnet** [huntabyte/shadcn-svelte#418] [Help] [answered]
Hi 👋 

I'm trying to get a reference to the HTML button when using the `Button` component.

If I were to use a native `button` I'd simply `bind:this={buttonRef}` to get access to a reference to my HTML button.

I wasn't able to achieve this using the `Button` component, any ideas ?

**Comment by Stadly** [comment]
I have the exact same issue. Is this possible, @huntabyte?

**Reply by shyakadavis** [reply]
Hi;

Thought this was implemented in https://github.com/huntabyte/bits-ui/pull/244, was it not?

**Reply by Stadly** [reply]
With `<Button bind:el={buttonRef}>`, `buttonRef` is undefined.

Looking at [bits-ui](https://github.com/huntabyte/bits-ui/blob/e966830792603d54f573ac682fd02b6ede039690/packages/bits-ui/src/lib/bits/button/components/button.svelte#L20), `bind:this={el}` is only used when `builders && builders.length`. Maybe that's the problem?

**Reply by Stadly** [reply]
I've created a PR to fix it: https://github.com/huntabyte/bits-ui/pull/473

**Comment by ttytm** [comment]
Again not possible with the `next` version?

**Reply by tmarnet** [reply, marked as answer]
It is possible using `bind:ref`, see https://next.bits-ui.com/docs/ref

**Reply by ttytm** [reply]
Thanks for sharing the reference.

Though I cannot use it with the default shadcn-svelte@next components unless I remove the fallback value `ref = $bindable(null)` -> `ref = $bindable()`.

If i keep it and try to use:
```svelte
	let searchElem = $state<HTMLInputElement>();
...
	<Input
		bind:ref={searchElem}
```
I'm getting:
```
Uncaught (in promise) Svelte error: props_invalid_value
Cannot do `bind:ref={undefined}` when `ref` has a fallback value
```
Works fine with the change mentioned above.

Should it be that way?

**Reply by shyakadavis** [reply]
@ttytm you need to provide a `null` default value.

**Reply by ttytm** [reply]
@shyakadavis so I need to do below? Which would work when i keep the component as it is.
```svelte
let searchElem = $state<HTMLInputElement | null>(null);
```

If I don't bother you. Why do the components use `ref = $bindable(null)` instead of `ref = $bindable()` in the first place?

**Reply by shyakadavis** [reply]
No worries. This should hold the answer, (if not, follow the entire thread, they go over it): https://github.com/huntabyte/shadcn-svelte/issues/1519#issuecomment-2508825393
